skrifa/src/color/mod.rs

```diff
@@ -0,0 +1,127 @@
+//! Color outline support.
+
+// Remove me when code is filled in
+#![allow(unused_variables, dead_code)]
+
+mod instance;
+
+use core::ops::Range;
+
+use crate::{instance::LocationRef, metrics::BoundingBox};
+use read_fonts::{tables::colr, types::GlyphId, ReadError, TableProvider};
+
+use instance::{resolve_paint, PaintId};
+
+/// Interface for receiving a sequence of commands that represent a flattened
+/// paint graph in a color outline.
+// Placeholder
+pub trait ColorPainter {}
+
+/// Affine transformation matrix.
+#[derive(Copy, Clone, Debug)]
+pub struct Transform {
+    xx: f32,
+    yx: f32,
+    xy: f32,
+    yy: f32,
+    dx: f32,
+    dy: f32,
+}
+
+/// Reference to a paint graph that represents a color glyph outline.
+#[derive(Clone)]
+pub struct ColorOutline<'a> {
+    colr: colr::Colr<'a>,
+    glyph_id: GlyphId,
+    kind: ColorOutlineKind<'a>,
+}
+
+impl<'a> ColorOutline<'a> {
+    /// Returns the glyph identifier that was used to retrieve this outline.
+    pub fn glyph_id(&self) -> GlyphId {
+        self.glyph_id
+    }
+
+    /// Evaluates the paint graph at the specified location in variation space
+    /// and returns the bounding box for the full scene.
+    ///
+    /// The `glyph_bounds` closure will be invoked for each clip node in the
+    /// graph and should return the bounding box for the given glyph, location
+    /// and transform.
+    pub fn bounding_box(
+        &self,
+        location: impl Into<LocationRef<'a>>,
+        glyph_bounds: impl FnMut(GlyphId, &LocationRef, Transform) -> Option<BoundingBox>,
+    ) -> Option<BoundingBox> {
+        None
+    }
+
+    /// Evaluates the paint graph at the specified location in variation space
+    /// and emits the results to the given painter.
+    pub fn paint(
+        &self,
+        location: impl Into<LocationRef<'a>>,
+        painter: &mut impl ColorPainter,
+    ) -> Result<(), ReadError> {
+        let instance = instance::ColrInstance::new(self.colr.clone(), location.into().coords());
+        match &self.kind {
```

Comment:
I am aware this is a draft - but how where does `self.kind` get initialized here - that would be a `ColorOutlineKind` retrieved for `self.glyph_id`, right?

Reply:
The kind field is initialized in `ColorOutlineCollection::get`. Yes, it is the outline for the associated glyph_id. I added that field (and accessor) to support iteration since the table can be sparse. It might make sense to remove that and change the iterator to yield a tuple `(GlyphId, ColorOutline)`